docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,207 @@
+  * Snaps: Server--staged to `$SNAP_DATA/secrets/edgex-redis/redis5-password` and injected via command line. (Marginal.). Clients--direct consumption from secret store. (Compliant.)
+
+- Kong authentication tokens
+  * Docker: System of reference is unencrypted Postgres database. (Non-compliant.)
+  * Snaps: System of reference is unencrypted Postgres database. (Non-compliant.)
+
+## Decision
+
+### Creation of secrets
+
+Management of hardware-bound secrets is platform-specific
+and out-of-scope for the EdgeX framework.
+EdgeX open source will contain only the necessary hooks to integrate
+platform-specific functionality.
+
+For software-managed secrets, the
+[_system of referece_](https://www.dqglossary.com/record%20of%20reference.html)
+of secrets in EdgeX is the EdgeX secret store.
+The EdgeX secret store provides for encryption of secrets at rest.
+This term means that if a secret is replicated,
+the EdgeX secret store is the authoritative source of truth of the secret.
+Whenever possible, the EdgeX secret store should also be the
+[_record of origin_](https://www.dqglossary.com/record%20of%20origin.html)
+of a secret as well.
+This means creating secrets inside inside of the EdgeX secret store
+is preferrable to importing an externally-created secret into the secret store.
+This can often be done for framework-managed secrets,
+but not possible for user-managed secrets.
+
+### Distribution and consumption of secrets
+
+#### Prohibited practices
+
+Use of hard-coded secrets is an instance of
+[CWE-798: Use of hard-coded credentials](https://cwe.mitre.org/data/definitions/798.html)
+and is not allowed.
+A hard-coded secret is a secret that is the same across multiple EdgeX instances.
+Hard-coded secrets make devices susceptible to BORE (break-once-run-everywhere) attacks,
+where collections of machines can compromised by a single replicated secret.
+Specific cases where this is likely to come up are:
+
+- Secrets embedded in source control
+
+  EdgeX is an open-source project.
+  Any secret that is present in an EdgeX repository is public to the world,
+  and therefore not a secret, by definition.
+  Configuration files, such as .toml files, .json files, .yaml files
+  (including `docker-compose.yml`) are specific instances of this practice.
+
+- Secrets embedded in binaries
+
+  Binaries are usually not protected against confidentiality threats,
+  and binaries can be easily reverse-engineered to find any secrets therein.
+  Binaries included compile executables as well as Docker images.
+
+#### Recommended practices
+
+1. Direct consumption from process-to-process interaction with secret store
+
+   This approach is only possible for components that have native support for
+   [Hashicorp Vault](https://www.vaultproject.io/).
+   This includes any EdgeX service that links to go-mod-secrets.
+
+2. Dynamic injection of secret into process environment space
```

Comment:
if I understand this consul-env would be a valid example here but I feel like we're missing a 2b which is consul-template?

Reply:
I added `envconsul` as an example.  Moreover, I added `consul-template` as the 3rd option, moved the 3rd option to the 4th option, and remove the original 4th option.  (I believe that `consul-template` can replace the 4th option for redis.)